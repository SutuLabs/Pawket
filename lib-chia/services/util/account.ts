import { AccountEntity, CustomCat, TokenInfo } from "../../models/account";
import receive, { TokenPuzzleDetail, TokenPuzzleObserver } from "../crypto/receive";

export const DEFAULT_ADDRESS_RETRIEVAL_COUNT = 1;

export async function getAccountAddressDetails(
  account: AccountEntity,
  cats: CustomCat[],
  tokenInfo: TokenInfo,
  prefix: string,
  symbol: string,
  maxId: number | undefined = undefined,
  catModName: "cat_v1" | "cat_v2" = "cat_v2"
): Promise<TokenPuzzleDetail[]> {
  if (typeof maxId !== "number" || maxId <= 0) maxId = account.addressRetrievalCount;
  if (typeof maxId !== "number" || maxId <= 0) DEFAULT_ADDRESS_RETRIEVAL_COUNT;
  if (!account.key.privateKey) throw new Error("Private key cannot empty for non-pk account to get account addresses.");

  if (account.addressGenerated == maxId) {
    return account.addressPuzzles;
  }

  if (account.type == "PublicKey") throw new Error("Public Key accound cannot provide puzzle detail");

  account.addressPuzzles = await receive.getAssetsRequestDetail(
    account.key.privateKey,
    0,
    maxId,
    cats,
    tokenInfo,
    prefix,
    symbol,
    catModName
  );
  account.addressGenerated = maxId;
  return account.addressPuzzles;
}

export async function getAccountPuzzleObservers(
  account: AccountEntity,
  cats: CustomCat[],
  tokenInfo: TokenInfo,
  prefix: string,
  symbol: string,
  maxId: number | undefined = undefined,
  catModName: "cat_v1" | "cat_v2" = "cat_v2"
): Promise<TokenPuzzleObserver[]> {
  if (typeof maxId !== "number" || maxId <= 0) maxId = account.addressRetrievalCount;
  if (typeof maxId !== "number" || maxId <= 0) DEFAULT_ADDRESS_RETRIEVAL_COUNT;

  if (account.addressGenerated == maxId) {
    return account.type == "PublicKey" ? account.observePuzzles ?? [] : account.addressPuzzles;
  }

  if (account.type == "PublicKey") {
    if (!account.key.publicKey) throw new Error("public key cannot empty for pk account to get account addresses.");
    account.observePuzzles = await receive.getAssetsRequestObserver(
      account.key.publicKey,
      0,
      maxId,
      cats,
      tokenInfo,
      prefix,
      symbol,
      catModName
    );
    account.addressGenerated = maxId;
    return account.observePuzzles;
  } else {
    if (!account.key.privateKey) throw new Error("Private key cannot empty for non-pk account to get account addresses.");
    account.addressPuzzles = await receive.getAssetsRequestDetail(
      account.key.privateKey,
      0,
      maxId,
      cats,
      tokenInfo,
      prefix,
      symbol,
      catModName
    );
    account.addressGenerated = maxId;
    return account.addressPuzzles;
  }
}
