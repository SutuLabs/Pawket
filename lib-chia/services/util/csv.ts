export function csvToArray(text: string): string[][] {
  // https://stackoverflow.com/a/8497474
  // Return array of string values, or NULL if CSV string not well formed.
  function CsvLinetoArray(text: string) {
    const re_valid =
      /^\s*(?:'[^'\\]*(?:\\[\S\s][^'\\]*)*'|"[^"\\]*(?:\\[\S\s][^"\\]*)*"|[^,'"\s\\]*(?:\s+[^,'"\s\\]+)*)\s*(?:,\s*(?:'[^'\\]*(?:\\[\S\s][^'\\]*)*'|"[^"\\]*(?:\\[\S\s][^"\\]*)*"|[^,'"\s\\]*(?:\s+[^,'"\s\\]+)*)\s*)*$/;
    const re_value =
      /(?!\s*$)\s*(?:'([^'\\]*(?:\\[\S\s][^'\\]*)*)'|"([^"\\]*(?:\\[\S\s][^"\\]*)*)"|([^,'"\s\\]*(?:\s+[^,'"\s\\]+)*))\s*(?:,|$)/g;
    // Return NULL if input string is not well formed CSV string.
    if (!re_valid.test(text)) return null;
    const a = []; // Initialize array to receive values.
    text.replace(
      re_value, // "Walk" the string using replace with callback.
      function (m0, m1, m2, m3) {
        // Remove backslash from \' in single quoted values.
        if (m1 !== undefined) a.push(m1.replace(/\\'/g, "'"));
        // Remove backslash from \" in double quoted values.
        else if (m2 !== undefined) a.push(m2.replace(/\\"/g, '"'));
        else if (m3 !== undefined) a.push(m3);
        return ""; // Return empty string.
      }
    );
    // Handle special case of empty last value.
    if (/,\s*$/.test(text)) a.push("");
    return a;
  }

  const lines = text.split("\n");
  const ops = [];
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (line.trim() == "") continue;

    const carr = CsvLinetoArray(line);
    if (carr == null) {
      // console.error("malformat of csv: ", line);
      throw new Error("malformat of csv: " + line);
    }

    ops.push(carr);
  }
  return ops;
}
