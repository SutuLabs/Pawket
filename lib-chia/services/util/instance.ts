import { ModuleInstance } from "@chiamine/bls-signatures";
import loadBls from "@chiamine/bls-signatures";
import * as clvm_tools from "clvm_tools";

class InstanceSingleton {
  private static _instance: InstanceSingleton = new InstanceSingleton();

  BLS?: ModuleInstance;

  constructor() {
    if (InstanceSingleton._instance) {
      throw new Error("Error: Instantiation failed: Use InstanceSingleton.getInstance() instead of new.");
    }
    InstanceSingleton._instance = this;
  }

  public static getInstance(): InstanceSingleton {
    return InstanceSingleton._instance;
  }

  public async init(): Promise<void> {
    this.BLS = await loadBls();
    await clvm_tools.initialize();
  }
}

export const Instance = InstanceSingleton.getInstance();
