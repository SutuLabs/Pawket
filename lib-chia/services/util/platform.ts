export function isIos(): boolean {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  if ((<any>window).cordova && (<any>window).cordova.platformId == "ios") return true;
  return false;
}
