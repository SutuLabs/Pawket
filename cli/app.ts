import "dotenv/config";
import express from "express";
import puzzle from "../lib-chia/services/crypto/puzzle";
import { Instance } from "../lib-chia/services/util/instance";
import { assemble } from "clvm_tools/clvm_tools/binutils";
import {
  analyzeCoin,
  convertUncurriedPuzzle,
  getModsPath,
  parseBlock,
  parseCoin,
  sexpAssemble,
  simplifyPuzzle,
  uncurryPuzzle,
} from "../lib-chia/services/coin/analyzer";
import { Hex0x, prefix0x } from "../lib-chia/services/coin/condition";
import { encodeOffer } from "../lib-chia/services/offer/encoding";
import { generateMintCnsOffer } from "../lib-chia/services/offer/cns";
import { OriginCoin, signSpendBundle } from "../lib-chia/services/spendbundle";
import transfer, { SymbolCoins, TransferTarget } from "../lib-chia/services/transfer/transfer";
import receive, { TokenPuzzleDetail } from "../lib-chia/services/crypto/receive";
import utility from "../lib-chia/services/crypto/utility";
import { NetworkContext } from "../lib-chia/services/coin/coinUtility";
import { getLineageProofPuzzle } from "../lib-chia/services/transfer/call";
import { CnsMetadataValues } from "../lib-chia/models/nft";
import { CustomCat } from "../lib-chia/models/account";
import { CoinItem } from "../lib-chia/models/wallet";
import fetch from "cross-fetch";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
(BigInt.prototype as any).toJSON = function () {
  return this.toString();
};
global.fetch = fetch;
const app: express.Express = express();
app.use(express.json({ limit: "3mb" }));
app.use(express.urlencoded({ extended: true }));
const defaultRpcUrl = "https://walletapi.chiabee.net/";

interface ParseBlockRequest {
  ref_list?: string[];
  generator: string;
}

interface ParsePuzzleRequest {
  puzzle: string;
}

interface AnalyzeTxRequest {
  coin_name: string;
  puzzle: string;
  solution: string;
  amount: number;
  coin_parent: string;
  puzzle_hash: string;
}

interface CnsOfferRequest {
  targetAddress: string;
  changeAddress: string;
  price: number;
  fee: number;
  metadata: CnsMetadataValues;
  coin: OriginCoin;
  privateKey: string;
  puzzleHash: string;
  puzzleText: string;
  royaltyAddress: string;
  royaltyPercentage: number;
  chainId?: string;
  symbol?: string;
  prefix?: string;
  nonce?: string; //test only
  intermediateKey?: string; //test only
  rpcUrl?: string;
}

interface TransferTargetOrigin {
  assetId: string;
  address: string;
  amount: bigint;
  memos?: string[];
}

interface BatchSendRequest {
  privateKey: string;
  transferTarget: TransferTargetOrigin[];
  changeAddress?: string;
  fee?: number;
  chainId?: string;
  symbol?: string;
  prefix?: string;
  availcoins?: SymbolCoins;
  rpcUrl?: string;
}

interface PuzzleRequest {
  method?: "ToAddress" | string;
  parameters?: string[];
}

Instance.init().then(() => {
  app.get("/version", async (_req: express.Request, res: express.Response) => {
    res.send(JSON.stringify({ version: "0.1" }));
  });

  app.post("/parse_block", async (req: express.Request, res: express.Response) => {
    try {
      if (process.env.SHOW_LOG) {
        console.time("parse_block");
      }
      const r = req.body as ParseBlockRequest;
      const bg = await parseBlock(r.generator, r.ref_list);
      const prog = sexpAssemble(bg);
      const argarr = await Promise.all(Array.from(prog.first().as_iter()).map(async (_) => await parseCoin(_)));

      if (process.env.SHOW_LOG) {
        console.log(`generated ${argarr.length} coins`);
      }
      // res.send(JSON.stringify(argarr, null, 4))
      res.send(JSON.stringify(argarr));
    } catch (err) {
      console.warn(err);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      res.status(500).send(JSON.stringify({ success: false, error: (<any>err).message }));
    } finally {
      if (process.env.SHOW_LOG) {
        console.timeEnd("parse_block");
      }
    }
  });

  app.post("/parse_puzzle", async (req: express.Request, res: express.Response) => {
    try {
      const r = req.body as ParsePuzzleRequest;
      const puz = await puzzle.disassemblePuzzle(r.puzzle);
      const decPuzzle = await simplifyPuzzle(assemble(puz));
      res.send(JSON.stringify(decPuzzle));
    } catch (err) {
      console.warn(err);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      res.status(500).send(JSON.stringify({ success: false, error: (<any>err).message }));
    }
  });

  app.post("/analyze_tx", async (req: express.Request, res: express.Response) => {
    let r: AnalyzeTxRequest | null = null;
    try {
      r = req.body as AnalyzeTxRequest;
      // console.log(`${JSON.stringify(r)},`);
      const coin = { amount: BigInt(r.amount), parent_coin_info: r.coin_parent as Hex0x, puzzle_hash: r.puzzle_hash as Hex0x };

      const uncPuzzle = await uncurryPuzzle(sexpAssemble(r.puzzle), r.puzzle);
      const decPuzzle = convertUncurriedPuzzle(uncPuzzle);
      const mods = getModsPath(decPuzzle);
      const analysis = await analyzeCoin(mods, uncPuzzle, coin, r.solution);

      res.send(
        JSON.stringify({
          coin_name: r.coin_name,
          parsed_puzzle: decPuzzle,
          mods,
          analysis: analysis,
        })
      );
    } catch (err) {
      console.warn(err);
      if (r) console.log(`${JSON.stringify(r)},`);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      res.status(500).send(JSON.stringify({ success: false, error: (<any>err).message }));
    }
  });

  app.post("/batch_send", async (req: express.Request, res: express.Response) => {
    let r: BatchSendRequest | null = null;
    try {
      r = req.body as BatchSendRequest;
      r.chainId = r.chainId || "ccd5bb71183532bff220ba46c268991a3ff07eb358e8255a65c30a2dce0e5fbb";
      r.prefix = r.prefix || "xch";
      r.symbol = r.symbol || "XCH";
      const rpcUrl = r.rpcUrl || defaultRpcUrl;

      const sk = Instance.BLS?.PrivateKey.from_bytes(utility.fromHexString(r.privateKey), false);
      if (!sk) {
        res.status(400).send(JSON.stringify({ success: false, error: "private key cannot be parsed" }));
        return;
      }

      const catList: CustomCat[] = [];
      const tokenNames: string[] = [];
      const tgts = r.transferTarget.map(
        (_) =>
          <TransferTarget>{
            symbol: _.assetId,
            address: prefix0x(puzzle.getPuzzleHashFromAddress(_.address)),
            amount: BigInt(_.amount),
            memos: _.memos,
          }
      );

      for (const tgt of tgts) {
        if (!catList.find((_) => _.id == tgt.symbol)) {
          catList.push({ name: tgt.symbol, id: tgt.symbol });
          tokenNames.push(tgt.symbol);
        }
      }
      const observers = await receive.getAssetsRequestDetail(r.privateKey, 0, 12, catList, {}, r.prefix, r.symbol, "cat_v2");
      const change_hex = prefix0x(
        r.changeAddress ? puzzle.getPuzzleHashFromAddress(r.changeAddress) : observers[0].puzzles[1].hash
      );

      if (change_hex.length != 66) {
        res.status(400).send(JSON.stringify({ success: false, error: "change address cannot be parsed or found." }));
        return;
      }

      let availcoins = r.availcoins;
      if (!availcoins) {
        const coins = (await receive.getActivities(observers, false, rpcUrl))
          .filter((_) => _.coin)
          .map((_) => _.coin as CoinItem)
          .map((_) => ({
            amount: BigInt(_.amount),
            parent_coin_info: _.parentCoinInfo,
            puzzle_hash: _.puzzleHash,
          }));
        availcoins = tokenNames
          .map((symbol) => {
            const tgtpuzs = observers.filter((_) => _.symbol == symbol)[0].puzzles.map((_) => prefix0x(_.hash));
            return { symbol, coins: coins.filter((_) => tgtpuzs.findIndex((p) => p == _.puzzle_hash) > -1) };
          })
          .reduce((a, c) => ({ ...a, [c.symbol]: c.coins }), {});
      }

      const plan = transfer.generateSpendPlan(availcoins, tgts, change_hex, BigInt(r.fee || 0), r.symbol);
      const net: NetworkContext = {
        chainId: r.chainId,
        prefix: r.prefix,
        symbol: r.symbol,
        api: (_) => getLineageProofPuzzle(_, rpcUrl),
      };
      const ubundle = await transfer.generateSpendBundleIncludingCat(plan, observers, [], net);
      const bundle = await signSpendBundle(ubundle, observers, net);

      res.send(
        JSON.stringify({
          spend_bundle: bundle,
        })
      );
    } catch (err) {
      console.warn(err);
      if (r) console.log(`${JSON.stringify(r)},`);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      res.status(500).send(JSON.stringify({ success: false, error: (<any>err).message }));
    }
  });

  app.post("/cns_offer", async (req: express.Request, res: express.Response) => {
    let r: CnsOfferRequest | null = null;
    try {
      r = req.body as CnsOfferRequest;
      if (!r.metadata) {
        res.status(400).send(JSON.stringify({ success: false, error: "metadata cannot be empty" }));
        return;
      }
      if (!r.metadata.bindings) r.metadata.bindings = {};

      // console.log(`${JSON.stringify(r)},`);

      r.chainId = r.chainId || "ccd5bb71183532bff220ba46c268991a3ff07eb358e8255a65c30a2dce0e5fbb";
      r.prefix = r.prefix || "xch";
      r.symbol = r.symbol || "XCH";
      const rpcUrl = r.rpcUrl || defaultRpcUrl;
      if (r.metadata.bindings.address?.startsWith("xch1"))
        r.metadata.bindings.address = puzzle.getPuzzleHashFromAddress(r.metadata.bindings.address);

      const availcoinsForMaker: SymbolCoins = {
        [r.symbol]: [
          {
            amount: 1n,
            parent_coin_info: r.coin.parent_coin_info,
            puzzle_hash: r.coin.puzzle_hash,
          },
        ],
      };
      const sk = Instance.BLS?.PrivateKey.from_bytes(utility.fromHexString(r.privateKey), false);
      if (!sk) {
        res.status(400).send(JSON.stringify({ success: false, error: "private key cannot be parsed" }));
        return;
      }

      const pubkey = utility.toHexString(sk.get_g1().serialize());
      const synPubKey = prefix0x(await puzzle.getSyntheticKey(pubkey));
      const tokenPuzzles: TokenPuzzleDetail[] = [
        {
          symbol: r.symbol,
          puzzles: [
            {
              privateKey: sk,
              synPubKey,
              puzzle: r.puzzleText,
              hash: r.puzzleHash,
              address: "",
            },
          ],
        },
      ];
      const royaltyAddressHex = puzzle.getPuzzleHashFromAddress(r.royaltyAddress);
      const net: NetworkContext = {
        chainId: r.chainId,
        prefix: r.prefix,
        symbol: r.symbol,
        api: (_) => getLineageProofPuzzle(_, rpcUrl),
      };
      const uofferBundle = await generateMintCnsOffer(
        r.targetAddress,
        r.changeAddress,
        BigInt(r.price),
        BigInt(r.fee),
        r.metadata,
        availcoinsForMaker,
        tokenPuzzles,
        royaltyAddressHex,
        r.royaltyPercentage,
        net,
        r.nonce,
        r.intermediateKey
      );
      const offerBundle = await signSpendBundle(uofferBundle, tokenPuzzles, net.chainId);
      const offer = await encodeOffer(offerBundle, 6);

      res.send(
        JSON.stringify({
          offer,
        })
      );
    } catch (err) {
      console.warn(err);
      if (r) console.log(`${JSON.stringify(r)},`);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      res.status(500).send(JSON.stringify({ success: false, error: (<any>err).message }));
    }
  });

  app.post("/puzzle", async (req: express.Request, res: express.Response) => {
    let r: PuzzleRequest | null = null;
    try {
      r = req.body as PuzzleRequest;

      if (r.method == "ToAddress" && r.parameters?.[0]) {
        res.send(
          JSON.stringify({
            address: puzzle.getAddressFromPuzzleHash(r.parameters?.[0], "xch"),
          })
        );
      } else {
        res.status(400).send(JSON.stringify({ success: false, error: "unrecognized method" }));
      }
    } catch (err) {
      console.warn(err);
      if (r) console.log(`${JSON.stringify(r)},`);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      res.status(500).send(JSON.stringify({ success: false, error: (<any>err).message }));
    }
  });
});

export default app;
