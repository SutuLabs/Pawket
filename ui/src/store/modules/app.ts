import { NotificationProgrammatic as Notification } from "buefy";
import store from "@/store";
import { ModuleInstance } from "@chiamine/bls-signatures";
import TestRunner from "../../services/selftest/runner";
import { tc } from "../../i18n/i18n";
import { Instance } from "../../../../lib-chia/services/util/instance";

export type SelfTestStatus = "Checking" | "Passed" | "Failed";
export interface IAppState {
  bls?: ModuleInstance;
  clvmInitialized: boolean;
  debug: boolean;
  selfTestStatus: SelfTestStatus;
  selfTestError: string;
}

store.registerModule<IAppState>("app", {
  state() {
    return {
      clvmInitialized: false,
      debug: false,
      bls: undefined,
      selfTestStatus: "Checking",
      selfTestError: "",
    };
  },
  actions: {
    async initializeClvm({ state }) {
      if (state.clvmInitialized) return;
      await Instance.init();
      state.clvmInitialized = true;
    },

    async initializeBls({ state }) {
      if (state.bls) return;
      await Instance.init();
      state.bls = Instance.BLS;
    },

    async selfTest({ state, dispatch }) {
      await dispatch("initializeBls");

      if (state.selfTestStatus == "Checking") {
        await TestRunner.selfTest();
      }
    },

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    copy({ state }, copyText: string) {
      const textArea = document.createElement("textarea");
      textArea.value = copyText;

      textArea.style.top = "0";
      textArea.style.left = "0";
      textArea.style.position = "fixed";

      document.body.appendChild(textArea);
      textArea.focus();
      textArea.select();

      try {
        document.execCommand("copy");
        Notification.open({
          message: tc("app.message.notification.copySuccess"),
          type: "is-primary",
        });
      } catch (err) {
        Notification.open({
          message: tc("app.message.notification.copyFailed"),
          type: "is-danger",
        });
      }

      document.body.removeChild(textArea);
    },
  },
});
