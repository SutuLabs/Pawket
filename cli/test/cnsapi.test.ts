import { Instance } from "../../lib-chia/services/util/instance";
import request from "supertest";
import app from "../app";
import case1 from "./cases/cns_case1.json";

beforeAll(async () => {
  await Instance.init();
});

jest.setTimeout(30000);

describe("CNS", () => {
  test("offer generation", async () => {
    const resp = await request(app).post("/cns_offer").send(case1);
    expect(resp.status).toBe(200);
    expect(JSON.parse(resp.text)).toMatchSnapshot("response");
  });
});
