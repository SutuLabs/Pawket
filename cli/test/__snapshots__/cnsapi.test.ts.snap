// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CNS offer generation: response 1`] = `
Object {
  "offer": "offer1qqr83wcuu2rykcmqvps0t0y3e0cd2glpf5lvwd3xums25u8d2t2cw0lykgegfnu9wtv8h4pl78cmv2gdjag2al0k7kjvf5tm4waa6f65jel3detudlffu99malhfh7gvzrqqsma8pmmnjm4scyda5d883wk8nkmpzt2ryf2k3864e6f8jkfr8veyxuhmt6y57t7lsge570g34zaxv5mmsfxuze9rqkpn838qy6mdsqjtmlctmeljs40jp3yr3y8k9h0k3ndpu464vak92knm44edhxuvcm7fmxpmv2ep4gd494ewtan2n6v7hhphd4lupdw5w8ufndtleetr2t48h2hkurlpfa3aqjx5rrgmk3syg65yk9q9yvjsry0h9ajracpjysse88lnyq3avwpjhz8ekqzrz863gx7vppleur4lxtl7nqey7uys8ralfju36vg6whw0a5ydrm9jlacftnex5vp4cyqtrkhf2asgjwe7utmfhv74tktw5fu7equ94w2595l703cj6lt6uaz32g2rwzjae5feygwymhk38x3dzwdzl7phersgj2mp0490xg9dullhu2phhfeemymd6jtjma77glzz07zynt9c6hk7g96mdtn07h6q4r6zqhzxl7rc6qxp38gej2w6yjqdrxdfuyanxp2w7q93d7xwpuhlxr6evtnkpv80kahsx5678w47f2n9eynwklsclrrnyne7qn930f0l7ajef9q5z3d7n6uhylnwffykazj6f9hfzkt3dfqkuhjewf02j5nxwfrxu6jjf6f9uln75elx5m4j0ec4avj6txg6ykd3hx0fv526gqede3v9pdnsqfm6dc0hlhc7e2cdmxxa2wfvxcu933rtss58lxd8ru4dz4nvltlh75mwa8lsxpfs4d9d8232f095cnftxwfh8r2nedxvmy5tfv9fxusnzv4chvjt2de88v622v4ck55v7we0xucj6v4fy5h4z29u55cnedf0tujjf060xgm5wh7mms7v3kwe6lydwjpuh55uw07pmscan5et6z6ayg4le8prxvpe9shucvw5hj4nrdqpveevwpcclk2e8rvua404l4dxru7067aglx32mak8048avn78jflttwvks7aq4gw2pu6jvr8ykjjv4dxy4n2v4a9yat92fqkn2nederxn22fteyhy6npf924jct3nxly5n242ft9rzj3dqdazjvw49ndaf0pjyw7n9t7qgshmxg9dtqhjltdyhqjt72p3842j7fen9adn7fen8y667wz44usz2r0y5jwgcp22axhdn97wl4u8m6ynjmfeg64kamp3ktf5hs3w4hew0tac7lt30p7sq5z4wy0xsc9khqzyffjkksjvwa98q9vhnwga39yuj2t6zhy3jsz72g9re3s4tev5uapgvgm0dje2pq2l7un6s47ktwafdkrnmf69ktxl8pe9l34mz9905weew29edp60vun576fgq7l73l89uysmfl6fn0vyj375w4al4fwtlzjp8fz0ml5zz0ne6nh9td7xmydwlzulj5zsk8w6v7l2dggxtvrjpjkhkw36pvur5zecapnsrl7nhydmdsrwjwjjfn7nm90pl8eydf9y7mpakaxr0g0g47usf265xrlc69v2m5rjpj56rzlcp8xhsq3kkr0nx89a2uga4uvaatxh7tg0d9xexpkl8pcdxkrhfqhkfhut2ledzl057gml7w2krhg2thnl27exnwk8rjhm4u9z79gm2z5lllchzqdfspxmpvdxzq3mqlv0sqn9v0ctprh3luzs2mmewhfplh3ursgkmw99t3srnaa84ldlmh680mw855hyuczu7zg4wdqn5zp2sfn37pszur70u7m0vd7pdhdnuhl24v8nknmlvy3t60ve083dajur3elel5lc49ulc02vrykpazzqrh5ykjswyuqeplhncgwk5r2s72je7t3ax9rx0p9wexkswz5uc5mj6l8z8nwqwmqc9v9y9ke5c6ah0jydzyg3p8d2a06n4mh9pk7t9k6njxj9qtr6e857khaal237uk2q5qc39fq7a7r4ar3qmfh8fv8dkl4m7jeef2azeet5rzfhl3dgl3ytd5wzvkjvzv3y0pme2yreq35809rcw728sau50pme2ye7t6ps9ltxcwp6esan4npm8txrkwkv8vavez6uu28saulzxfh8fxrhwjvp0ywmuzgcl8qsg6rhjzftr5w7dlfl8txrkwkv80vgw7gfchd07pvheflh6am3n67zag4paldxr7dfzl8ku7tnudpg2jd3dm2samd9q90lacqapqnrgdqtrcyzdfqgvjt0mlc99u0dwe0ne5yaf7ptn9lk75mlmtzvuszax0t7hjahnguzs779f3khuq7tsewh79retx49dmxhnw2mstkev8rma3mmhftk58yurhcyt6hfntu30l3gwdvmd0zwd9w3nt9j0h8dfjd4fehgcy33tjmk8rumm92y773t2xae8hvu0k8f357jw296g9adg7wchalm04chauw585c0aa6ka95nrucmp0hqyq8nqqygjsgznhz",
}
`;
